File: src/Algorithm/LeetCode/LeetCode_59_Spiral_Matrix_II/Solution.java
package Algorithm.LeetCode.LeetCode_59_Spiral_Matrix_II;

class Solution {
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];

        if (n == 0) {
            return matrix;
        }

        int top = 0;
        int bottom = n - 1;
        int left = 0;
        int right = n - 1;

        int i = 1;

        while (true) {
            for (int column = left; column <= right; column++) {
                matrix[top][column] = i++;
            }

            if (++top > bottom) {
                break;
            }

            for (int row = top; row <= bottom; row++) {
                matrix[row][right] = i++;
            }

            if (--right < left) {
                break;
            }

            for (int column = right; column >= left; column--) {
                matrix[bottom][column] = i++;
            }

            if (--bottom < top) {
                break;
            }

            for (int row = bottom; row >= top; row--) {
                matrix[row][left] = i++;
            }

            if (++left > right) {
                break;
            }
        }

        return matrix;
    }
}

File: src/Algorithm/LeetCode/LeetCode_59_Spiral_Matrix_II/Solution.py
from typing import List


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        matrix = [[0] * n for _ in range(n)]
        left, right = 0, n - 1
        top, bottom = 0, n - 1
        count = 0

        while True:
            for column in range(left, right + 1):
                count += 1
                matrix[top][column] = count

            top += 1
            if top > bottom:
                break

            for row in range(top, bottom + 1):
                count += 1
                matrix[row][right] = count

            right -= 1
            if right < left:
                break

            for column in range(right, left - 1, -1):
                count += 1
                matrix[bottom][column] = count

            bottom -= 1
            if bottom < top:
                break

            for row in range(bottom, top - 1, -1):
                count += 1
                matrix[row][left] = count

            left += 1
            if left > right:
                break

        return matrix
